### PR TITLE
Extra requirement support (extras_require)

### LINKED ISSUE #625: Pip-compile ignores extras_require when parsing setup.py
The documentation for python-wheel describes ways to define conditional dependencies: http://wheel.readthedocs.io/en/stable/index.html?highlight=extras_require#defining-conditional-dependencies

The most well-supported of these is to use the parameter `extras_require` to specify dependencies for specific environment markers. For example:

```
from setuptools import setup
setup(
    name='mypackage',
    extras_require={
        ':platform_python_implementation=="CPython"': ['python-cjson'],
    }
)
```

This allows us to install optimisations for CPython which are not available on other implementations of Python. This works correctly for pip/setuptools (e.g. `pip install .` or `python setup.py install`) but is ignored by pip-compile.

##### Environment Versions

1. OS Type (Linux/Ubuntu)
1. Python version: 2.7
1. pip version: 9.0.1
1. pip-tools version: 1.11.0

##### Steps to replicate

1. Create a package with a conditional dependency specified using `extras_require` in `setup.py`.
2. Run pip-compile on the package in an environment which meets the conditions, and in one that does not.

##### Expected result

The generated requirements.txt should contain the conditional packages for the environment which meets the conditions, and not otherwise.

##### Actual result

The conditional dependencies are not present in any requirements files.

### PR DESCRIPTION
**Pull request by orsinium**
**Changelog-friendly one-liner**: Add `--extra` to specify `extras_require` dependencies.

Supports setup.py and anything PEP-517 compatible. Tested for setup.py, setup.cfg, poetry, and flit.

Close #625.

##### Contributor checklist

- [x] Provided the tests for the changes.
- [x] Gave a clear one-line description in the PR (that the maintainers can add to CHANGELOG.md on release).
- [x] Assign the PR to an existing or new milestone for the target version (following [Semantic Versioning](https://blog.versioneye.com/2014/01/16/semantic-versioning/)).

### CODE REVIEW

**Comment by orsinium** [comment]
I hope, we'll be able to take it into #1362, it's closely related to #1356, they play nice together.

**Comment by atugushev** [comment]
> Ready for review. I hope, we'll be able to take it into #1362, it's closely related to #1356, they play nice together.

Ok, I've added it to `6.1.0`.

File: piptools/scripts/compile.py
Comment (atugushev) [inline piptools/scripts/compile.py:71]:
```suggestion
    "-E",
```

I'would like to suggest `-E` instead of `-e`, which clashes with `pip install -e`.

Comment (orsinium) [reply piptools/scripts/compile.py:71]: Good catch. Let's just drop the short key and leave only `--extras`

Comment (orsinium) [reply piptools/scripts/compile.py]: Fixed :+1:

**Comment by orsinium** [comment]
Is there anything that remains to do?

**Comment by atugushev** [comment]
Could you add a test for `pip-compile --extra=whatever requirements.in`?

**Comment by atugushev** [comment]
@orsinium IMO there should be an error.

**Comment by orsinium** [comment]
Done 👍🏿

File: tests/conftest.py
Comment (webknjaz) [inline tests/conftest.py:348]: Please add a docstring to improve the UX of the `pytest --fixtures` invocation.

File: tests/test_cli_compile.py
Comment (webknjaz) [inline tests/test_cli_compile.py:1742]: Is it intentional that these are duplicated in here and in the extras?

File: tests/test_cli_compile.py
Comment (webknjaz) [inline tests/test_cli_compile.py:1782]: I don't fully understand. Is "extra" supposed to only select the dists that are listed in extras but not the main runtime deps?

File: piptools/scripts/compile.py
Comment (webknjaz) [inline piptools/scripts/compile.py:74]: 1) `s/n/N`
2) If this is supposed to install only extras but not the default runtime deps, can this be renamed into `--extra-only`?
3) If it's supposed to merge those extras with the main runtime deps list, can we add an additional flag `--extras-only`/`--exclude-default-deps` or something?

File: tests/test_cli_compile.py
Comment (webknjaz) [inline tests/test_cli_compile.py:1757]: It'd be great if you could add docstrings to the tests so that it'd be clear what expectations they assume.

File: tests/test_cli_compile.py
Comment (orsinium) [reply tests/test_cli_compile.py:1782]: small-fake-e and small-fake-f are from another extra, the `test` one, we pass only `dev`.

File: tests/test_cli_compile.py
Comment (orsinium) [reply tests/test_cli_compile.py:1742]: Yes, this is how poetry works. Here is the specification, in `extras` only names.

File: piptools/scripts/compile.py
Comment (webknjaz) [reply piptools/scripts/compile.py]: It's actually a long-standing feature request in pip and I think there's been no opposition, just nobody had time to come up with an implementation. The main motivation was that people often supply test/dev/docs extras for different processes they have and some of them don't really require the dist itself to be installed. Of course, there's another camp saying that this sort of deps should be put elsewhere because they aren't necessary for the runtime and there should be env pins anyway.
But if pip-tools were to implement this flag, it'd facilitate a hybrid of these two approaches with the direct open-ended deps in the dist metadata configs and the pins existing in separate constraints files.
But yes, this could probably be brought up in a separate discussion.

**Review by webknjaz — APPROVED** [review]
I think it's merge-ready, any other comments can be addressed separately.

**Comment by atugushev** [comment]
I've run `pip-compile setup.py --extra=testing` on [setup.py](https://github.com/jazzband/pip-tools/blob/master/setup.cfg) and noticed `extra` markers.


<details>

 <summary>Output</summary>

```ini
#
# This file is autogenerated by pip-compile
# To update, run:
#
#    pip-compile --extra=testing setup.py
#
apipkg==1.5
    # via execnet
attrs==20.3.0
    # via pytest
click==7.1.2
    # via pip-tools (setup.py)
execnet==1.8.0
    # via pytest-xdist
iniconfig==1.1.1
    # via pytest
packaging==20.9
    # via pytest
pep517==0.10.0
    # via pip-tools (setup.py)
pluggy==0.13.1
    # via pytest
py==1.10.0
    # via
    #   pytest
    #   pytest-forked
pyparsing==2.4.7
    # via packaging
pytest-forked==1.3.0
    # via pytest-xdist
pytest-rerunfailures==9.1.1 ; extra == "testing"
    # via pip-tools (setup.py)
pytest-xdist==2.2.1 ; extra == "testing"
    # via pip-tools (setup.py)
pytest==6.2.2 ; extra == "testing"
    # via
    #   pip-tools (setup.py)
    #   pytest-forked
    #   pytest-rerunfailures
    #   pytest-xdist
toml==0.10.2
    # via
    #   pep517
    #   pytest

# The following packages are considered to be unsafe in a requirements file:
# pip
# setuptools
```

</details>


Note `pytest==6.2.2 ; extra == "testing"`.

I've tried to install the result `requirements.txt` and got the following "warnings":

```console
$ pip install -r requirements.txt
Ignoring pytest-rerunfailures: markers 'extra == "testing"' don't match your environment
Ignoring pytest-xdist: markers 'extra == "testing"' don't match your environment
Ignoring pytest: markers 'extra == "testing"' don't match your environment
...
```

This is FYI, I haven't researched it yet deeply.

**Comment by webknjaz** [comment]
> Note `pytest==6.2.2 ; extra == "testing"`.

Ouch... You can't use extras as markers. Only a restricted subset of vars/functions is available there.

The table of available markers is here: https://www.python.org/dev/peps/pep-0508/#environment-markers.

For `extra` it specifically says:
> An error except when defined by the context interpreting the specification.

And so I haven't heard of any tool implementing this. At least, I'm fairly sure that pip doesn't have any interface to specify extras in requirements files.

**Comment by atugushev** [comment]
@webknjaz thanks for the info! Perhaps, "extra" should be filtered out in [format_requirement](https://github.com/jazzband/pip-tools/blob/019516ae8a16de739370b8e5329580fcb8fa283c/piptools/utils.py#L117) somehow. Ideas?

File: piptools/scripts/compile.py
Comment (atugushev) [inline piptools/scripts/compile.py:409]: [InstallRequirement.match_markers(extras)](https://github.com/pypa/pip/blob/7ec0fa5142466f8402bb866a1dcfb8ca6ffdc66b/src/pip/_internal/req/req_install.py#L268) could be used instead of `req_check_markers()`. Looks like it works the same way under the hoods.
```suggestion
    constraints = [req for req in constraints if req.match_markers(extras)]
```

File: piptools/scripts/compile.py
Comment (orsinium) [reply piptools/scripts/compile.py]: I'm from that camp too. I have a separate environment for every tool. However, most of the tools still need dependencies: `pytest`  for sure, `mypy` for type-checking of third-party libs usage, `pylint` for inference-based tasks. In my projects, only `flake8` works nicely without knowing the project dependencies. And I keep a separate `requirements.txt` for it. Yep, makes the root of the project a bit messier but works. This is why I made [dephell](https://github.com/dephell/dephell/) a long time ago, to handle all this zoo :)

**Comment by orsinium** [comment]
Possible solutions I see for removing `extra` marker:

1. Hacking with regexp
2. Hacking through inspection of `packaging.Marker._marker`. It has an interesting structure of tuples in lists inside. If we assume that `extra` always comes at the top level, we can easily find and remove it. That means getting into a "private" method but `packaging` should be very stable and unlikely to change.
3. https://github.com/dephell/dephell_markers has `.remove` method. I made it a long time ago for dephell, its well well tested and works. Still, a third-party dependency for a small task.

**Comment by atugushev** [comment]
Approach №2 is fine by me 👍🏻

**Comment by orsinium** [comment]
I've implemented and integrated `drop_extras` function which removes `extra` from markers. Unit tests are included and all seems to be working fine. However, I can't cover one line somehow 🤔

**Comment by orsinium** [comment]
Huh. I've put a breakpoint there and it IS covered, for sure. I have a deja vu. Coverage.py seems to be quite flaky about branch coverage for if conditions. I definitely faced something like this in [deal](https://github.com/life4/deal) but I don't remember the exact solution. Probably, I've just rewritten the condition back then.

**Comment by orsinium** [comment]
I decided to hack it around here too 🙃

**Review by ssbarnea — APPROVED** [review]

**Comment by ssbarnea** [comment]
@atugushev @orsinium @webknjaz Is there anything blocking progress on this PR? I do find it very useful and a blocker for the new release.

**Comment by orsinium** [comment]
Nothing from my side. There is a new logic (https://github.com/jazzband/pip-tools/pull/1363/commits/3b63aa94370185094b6a827defcd2bc9b2551b22) that needs a review from maintainers.

**Comment by atugushev** [comment]
Hello @ssbarnea! Thanks for pinging. I'm going to look into this shortly.

**Comment by ssbarnea** [comment]
I should be the owe thanking you. That is indeed a very important feature, one that prevented further adoption from more people.

**Review by atugushev — APPROVED** [review]
@orsinium thanks for fixing the bug with `extras` 🙏🏻 The PR's all good!

**Comment by atugushev** [comment]
`pip-tools v6.1.0` is released 🚀

**Pull request merged**
